### ISSUE TITLE
Cannot connect to socket.io V3

### ISSUE THREAD

**Issue by aboda1986** [rikulo/socket.io-client-dart#130]
I try to connect with Socket io **V3** as the following code:

index.js

```js
var app = require('express')();
var http = require('http').createServer(app);
var io = require('socket.io')(http ,{
    cors: {
        origins: '*:*',
      methods: ["GET", "POST"],
   
      credentials: true
    }
  });

  var port = process.env.PORT || 3000;


  io.on('connection', function(socket){
    socket.on('chat message', function(msg){
      io.emit('chat message', msg);
      console.log(msg);
    });
  });
  

http.listen(port, () => {
  console.log('listening on *:' +port);
});
```

main.dart

```dart
import 'package:flutter/material.dart';

import 'package:socket_io_client/socket_io_client.dart' as IO;

void main() async {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
     
        primarySwatch: Colors.blue,
      
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  void initState() {
    super.initState();

    _connect();
  }

  void _connect() async {
    IO.Socket socket = IO.io('https://jeranapp.com', <String, dynamic>{
      'transports': ['websocket']
    });

    socket.on('connect', (_) {
      print('connect');
      socket.emit('chat message', 'from app wooow');
    });

    socket.on('data', (_) {
      print(_);
    });

    socket.on('connect_error', (data) => print(data));
    socket.on('connect_timeout', (data) => print(data));
    socket.on('connecting', (data) => print(data));
    socket.on('disconnect', (data) => print(data));
    socket.on('error', (data) => print(data));
    socket.on('reconnect', (data) => print(data));
    socket.on('reconnect_attempt', (data) => print(data));
    socket.on('reconnect_failed', (_) => print(_));
    socket.on('reconnect_error', (_) => print(_));

    socket.on('reconnecting', (_) => print(_));
    socket.on('ping', (_) => print(_));
    socket.on('pong', (_) => print(_));
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text(
              'You have pushed the button this many times:',
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () {
          _connect();
        },
        tooltip: 'Increment',
        child: Icon(Icons.add),
      ),
    );
  }
}
```

the result message in console are "ping timeout" and "transport close".

plugins is connected fine with  socketio v2.

kindly advise

**Comment by jumperchen** [comment]
Socket.io V3 is a BIG version, so it may change something internally; we are not yet planning to support it for this library.

**Comment by billbull21** [comment]
please,, fix this issue. i feel so sad cannot implement chat feature on my app. :(

**Comment by jumperchen** [comment]
@billbull21 Why can't you use Socket.io v2 for chat feature?

**Comment by jumperchen** [comment]
memo:

The migration guide - https://socket.io/docs/v3/migrating-from-2-x-to-3-0/
> TL;DR: due to several breaking changes, a v2 client will not be able to connect to a v3 server (and vice versa)

These two changes break this library to work with Socket.io V3
1. [Engine.IO protocol v4](https://github.com/socketio/engine.io-protocol#difference-between-v3-and-v4)
2. [Socket.IO protocol v5](https://github.com/socketio/socket.io-protocol#difference-between-v5-and-v4)

**Comment by billbull21** [comment]
> @billbull21 Why can't you use Socket.io v2 for chat feature?

my server use socket.io v3 for chat features. which is my friend was setting it. i cannot force my friend to use v2 cause the program was running. me, as a client just wanted to know how can i connect socket.io from server v3 to my flutter app using this plugin. what is the only way to use it is downgrade the socket io in the server side?

thanks in advance

**Comment by jumperchen** [comment]
@billbull21 Yes, downgrade the server side version is the fastest way, or welcome to contribute the code from V2 to V3. :)

**Comment by steve-cahn** [comment]
Any plans on supporting v3?

**Comment by jumperchen** [comment]
@steve-cahn Yes, it will be next big version for the `socket.io-client-dart`, but so far I don't have free time to migrate it from V2 to V3 (because we don't use V3 internally); would you like to contribute this for the two protocols changes mentioned in https://github.com/rikulo/socket.io-client-dart/issues/130#issuecomment-738522845 ?

**Comment by ali-tafakkori** [comment]
+1

**Comment by jumperchen** [comment]
Resolved in `2.0-beta` branch.

**Comment by jumperchen** [comment]
Hi all,

please try with `v2.0.0-beta.1` version for Socket.io V3 server.

**Comment by billbull21** [comment]
> Hi all,
> 
> please try with `v2.0.0-beta.1` version for Socket.io V3 server.

you are my savior.
thanks.

**Comment by WygorFonseca** [comment]
@jumperchen Can i contribute?

**Comment by jumperchen** [comment]
@WygorFonseca sure, but the socket.io V3 already supported in `2.0-beta` branch and released on **pub.dev** as well.
